tests/test_utils.py

```diff
@@ -0,0 +1,167 @@
+    assert nodes[0].parser.html == '<!DOCTYPE html><html><head><link href="http://"></head><body><div><script src="http://"></script></div></body></html>'
+
+    assert len(nodes[0].parser.css("head")) == 1
+    assert len(nodes[0].parser.css("body")) == 1
+
+
+@pytest.mark.parametrize(*_IMPL_PARAMETRIZER)
+def test_parse_fragment_head_and_body_no_html(impl: Impl):
+    html = '<head><link href="http://"></head><body><div><script src="http://"></script></div></body>'
+    nodes = impl.parse_fragment_fn(html)
+    assert len(nodes) == 2
+    assert nodes[0].tag == "head"
+    assert nodes[1].tag == "body"
+    assert nodes[0].html == '<head><link href="http://"></head>'
+    assert nodes[1].html == '<body><div><script src="http://"></script></div></body>'
+    assert nodes[0].parser.html == '<html><head><link href="http://"></head><body><div><script src="http://"></script></div></body></html>'
+
+    assert len(nodes[0].parser.css("head")) == 1
+    assert len(nodes[0].parser.css("body")) == 1
+
+
+@pytest.mark.parametrize(*_IMPL_PARAMETRIZER)
+def test_parse_fragment_head_no_html(impl: Impl):
+    html = '<head><link href="http://"></head>'
+    nodes = impl.parse_fragment_fn(html)
+    assert len(nodes) == 1
+    assert nodes[0].tag == "head"
+    assert nodes[0].html == '<head><link href="http://"></head>'
+    assert nodes[0].parser.html == '<html><head><link href="http://"></head></html>'
+
+    assert len(nodes[0].parser.css("head")) == 1
+    assert len(nodes[0].parser.css("body")) == 0
+
+
+@pytest.mark.parametrize(*_IMPL_PARAMETRIZER)
+def test_parse_fragment_body_no_html(impl: Impl):
+    html = '<body><div><script src="http://"></script></div></body>'
+    nodes = impl.parse_fragment_fn(html)
+    assert len(nodes) == 1
+    assert nodes[0].tag == "body"
+    assert nodes[0].html == '<body><div><script src="http://"></script></div></body>'
+    assert nodes[0].parser.html == '<html><body><div><script src="http://"></script></div></body></html>'
+
+    assert len(nodes[0].parser.css("head")) == 0
+    assert len(nodes[0].parser.css("body")) == 1
+
+
+@pytest.mark.parametrize(*_IMPL_PARAMETRIZER)
+def test_parse_fragment_fragment(impl: Impl):
+    html = '<link href="http://"><div><script src="http://"></script></div>'
+    nodes = impl.parse_fragment_fn(html)
+    assert len(nodes) == 2
+    assert nodes[0].tag == "link"
+    assert nodes[1].tag == "div"
+    assert nodes[0].html == '<link href="http://">'
+    assert nodes[1].html == '<div><script src="http://"></script></div>'
+
+    # NOTE: Ideally the full HTML would NOT contain `<html>`, `<head>` and `<body>` in this case,
+    # but this is technical limitation of the parser.
+    # But as long as user serializes fragment nodes by as `Node.html`, they should be fine.
+    assert nodes[0].parser.html == '<html><head><link href="http://"></head><body><div><script src="http://"></script></div></body></html>'
```

Comment:
Also, one more caveat of this approach. The list of nodes returned from `parse_fragment` will correctly only contain those elements, as were defined at the root of the HTML string. However, if user goes to the parent HTMLParser to render the HTML, then the resulting HTMLwill contain also `<html>` and other.

But this could be also considered a feature. That one can still render the full HTML even if they have access to the nodes. So, dunno... 😄 🤷 I guess important is to be aware of this.